pyaerocom/io/read_aeronet_sunv3.py

```diff
@@ -124,91 +129,118 @@ def read_file(self, filename, vars_to_retrieve=None,
-                else:
-                    self.logger.warning("Variable {} not available in file {}"
-                                        .format(var, os.path.basename(filename)))
-            pl = None
-            for i, line in enumerate(in_file):
-                # process line
-                dummy_arr = line.split(self.COL_DELIM)
-
-                if pl is not None and len(dummy_arr) != len(pl):
-                    const.print_log.exception('Data line {} in {} is corrupt, '
-                                              'skipping...'.format(i, filename))
-                    continue
-                # copy the meta data (array of type string)
-                for var in self.META_NAMES_FILE:
-                    try:
-                        val = dummy_arr[col_index[var]]
-                    except IndexError as e:
-                        const.print_log.exception(repr(e))
-
-                    try:
-                        # e.g. lon, lat, altitude
-                        val = float(val)
-                    except Exception:
-                        pass
-                    data_out[var].append(val)
-
-                # This uses the numpy datestring64 functions that e.g. also
-                # support Months as a time step for timedelta
-                # Build a proper ISO 8601 UTC date string
-                day, month, year = dummy_arr[col_index['date']].split(':')
-                datestring = '-'.join([year, month, day])
-                datestring = 'T'.join([datestring, dummy_arr[col_index['time']]])
-                # NOTE JGLISS: parsing timezone offset was removed on 22/2/19
-                # since it is deprecated in recent numpy versions, for details
-                # see https://www.numpy.org/devdocs/reference/arrays.datetime.html#changes-with-numpy-1-11
-                # datestring = '+'.join([datestring, '00:00'])
-
-                data_out['dtime'].append(np.datetime64(datestring))
-
-                # TODO: remove elif if ensured that it works
-                for var, idx in vars_available.items():
-                    val = np.float_(dummy_arr[idx])
-                    if val == self.NAN_VAL:
-                        val = np.nan
-                    data_out[var].append(val)
-
-                pl = dummy_arr
+        # enable alternative reading of .gz files here to save space on the file system
+        suffix = pathlib.Path(filename).suffix
+        tmp_name = filename
+        if suffix == '.gz':
+            f_out = tempfile.NamedTemporaryFile(delete=False)
+            with gzip.open(filename, 'r') as f_in:
+                shutil.copyfileobj(f_in, f_out)
+            filename = f_out.name
+            f_out.close()
+
+        try:
+            with open(filename, 'rt') as in_file:
+                lines = in_file.readlines()
+        except UnicodeDecodeError:
+            with open(filename, 'rt', encoding='ISO-8859-1') as in_file:
+                lines = in_file.readlines()
```

Comment:
@avaldebe do you know a more elegant way of putting a context manager in a try block?
The goal would be to not read the entire file first. But the encoding error can occur anywhere, so one would need to change the encoding on the fly. 
Any suggestions (but don't think too much about it). 
But this is the 2nd obs network whose data files can have different encodings (the other is EEA)

Reply:
google suggested me this solutions:

[Check for BOM markers](https://unicodebook.readthedocs.io/guess_encoding.html#check-for-bom-markers)

``` python
from codecs import BOM_UTF8, BOM_UTF16_BE, BOM_UTF16_LE, BOM_UTF32_BE, BOM_UTF32_LE

BOMS = (
    (BOM_UTF8, "UTF-8"),
    (BOM_UTF32_BE, "UTF-32-BE"),
    (BOM_UTF32_LE, "UTF-32-LE"),
    (BOM_UTF16_BE, "UTF-16-BE"),
    (BOM_UTF16_LE, "UTF-16-LE"),
)

def check_bom(data):
    return [encoding for bom, encoding in BOMS if data.startswith(bom)]
```

[Is UTF-8?](https://unicodebook.readthedocs.io/guess_encoding.html#is-utf-8)

``` python
def isUTF8(data):
    try:
        data.decode('UTF-8')
    except UnicodeDecodeError:
        return False
    else:
        return True
```

For any of these to work, you would need to read the file as binary and decode on a second step.
In any case you need to read the file in one go.

Alas, I do not see a clean way to guess the encoding without reading the file.
Is this a problem big enough to depend on a 3rd party library?

Reply:
no I don't want some external library dependency for this. The BOM approach also reads the beginning of the file as does my code (the crash happens in the file header), so there's no real gain.
I therefore think I leave it as it is at least for now. Only the all points Aeronet file can be big, but not gig enough to really bother.